### ISSUE TITLE
 Propagation of options to dependencies

### ISSUE THREAD

**Issue by sbellus** [conan-io/conan#9690]
Let's have libraries M,A,B with following dependency graph
M -uses-> A -uses with option o=X-> B
M -uses with option o=Y-> B

How can I write recipe for M?
My observation is that conan uses only o=Y for M and also for A, even if A requests B to build with option o=X.
It results to build failure because A is not able to depend on B with option o=Y.

**Comment by memsharded** [comment]
Hi @sbellus 

I think you might be interested in this section of the docs: https://docs.conan.io/en/latest/versioning/introduction.html#version-and-configuration-conflicts

The issue is that 2 different binary variants of the same package cannot be part of the same dependency graph, as it would create linking conflicts (you cannot link the same library as shared and static in the same executable). These configuration conflicts are resolved by the top downstream consumer, in this case M, if M says the option=Y, then that option will be used and propagated to the whole graph.

Please let me know if that clarifies it, otherwise I suggest making the package names and options real, that might illustrate better a possible use case.

**Comment by sbellus** [comment]
It clarifies observed behavior. From my point of view it is bad behavior.
I see in all examples shared, static. It is not about it at all.

package A produces A.so (always shared library)
package B produces B.a (always static library)
package M produces M.so (always shared library)

There is no linkage issue because A.so can link different variant of B.a as M.so and it is OK.

We use option to modify functionality of B. Lets say option with_image_support. The B then exports additional headers.
package M does not want to use image support from B because it implements own image manipulation that conflicts with B.
BUT when M set this option to False conan propagates it also for B and headers are missing during B build.

Does conan support some syntax to keep option valid lets say in direct dependency?
E.g. self.options.Set("B:with_image_support", False, overall=False)?

**Comment by memsharded** [comment]
> There is no linkage issue because A.so can link different variant of B.a as M.so and it is OK.

Not in the general case. For example, headers inlining also produce binary changes in the consumers, which can lead to binary incompatibility errors. It is also very typical that the library itself contains other binary changes, beyond some extra headers, like new libraries. If not, you might consider just exporting those headers always, it will be harmless to the consumer if they don't include those headers. And if they implement different binary artifacts, then the general case is duplicity of symbols and linking errors. Even in the extreme case that the APIs are fully isolated, there still could be weird internal ODR violations and ABI incompatibilities. We have seen this case a few times already, M exe linking with A shared library and with B static library, and A shared library statically linking a slightly different version of B static library: byte alignment issues, memory issues, etc. This is the reason why Conan implement the current behavior, because it is the safest one in the general case.

If you still want to achieve dependencies isolation and duplication you have the following alternatives:

- Use the ``private`` specifier for dependency in Conan 1.X. This is not recommended in the general case, as it has some quirks. You will have 2 independent B packages in your graph, which can have different version and configuration, including options.
- Wait until Conan 2.0, where the dependency graph model will be more elaborated and allow modelling this use case better

**Comment by sbellus** [comment]
The private flag does not work.
I tried following combinations:
package A has self.requires("B/2.50.0", private=True)
package M has self.requires("B/2.52.0", private=True)
And
package A has self.requires("B/2.50.0", private=True)
package M has self.requires("B/2.52.0")

None of them works.

**Comment by memsharded** [comment]
Could you please elaborate how it fails? It would be great to have something to be able to reproduce, post the output with the error messages...

So far I have done this test with this setup that you are describing it fails, and seems to be working fine:

```python
def test_private_error():
    # https://github.com/conan-io/conan/issues/9690
    c = TestClient()
    c.save({"pkgb/conanfile.py": GenConanfile(),
            "pkga/conanfile.py": GenConanfile().with_require("pkgb/2.50.0", private=True),
            "pkgm/conanfile.py": GenConanfile().with_require("pkgb/2.52.0", private=True)
                                               .with_require("pkga/0.1")})
    c.run("create pkgb pkgb/2.50.0@")
    c.run("create pkgb pkgb/2.52.0@")
    c.run("create pkga pkga/0.1@")
    c.run("create pkgm pkgm/0.1@")
    assert "pkgb/2.50.0 from local cache - Cache" in c.out
    assert "pkgb/2.52.0 from local cache - Cache" in c.out
    assert "pkgb/2.50.0:5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9 - Skip" in c.out
    assert "pkgb/2.52.0:5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9 - Cache" in c.out

    c.run("create pkgm pkgm/0.1@ --build")
    assert "pkgb/2.50.0 from local cache - Cache"
    assert "pkgb/2.52.0 from local cache - Cache"
    assert "pkgb/2.50.0:5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9 - Skip" not in c.out
    assert "pkgb/2.52.0:5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9 - Build" in c.out
```

**Comment by sbellus** [comment]
Yes this is ok. conan builds 2 version of library. Is it possible to make test with following pseudo code?
```
def test_private_error():
    # https://github.com/conan-io/conan/issues/9690
    c = TestClient()
    c.save({"pkgb/conanfile.py": GenConanfile().with_option("o", ["X", "Y"]),
            "pkga/conanfile.py": GenConanfile().with_require("pkgb/2.50.0", private=True),
                                               .with_option("b:o", "X")
            "pkgm/conanfile.py": GenConanfile().with_require("pkgb/2.52.0", private=True)
                                               .with_option("b:o", "Y")
                                               .with_require("pkga/0.1")})
    c.run("create pkgm pkgm/0.1@")
    assert "o=X" in pkgb/2.50.0:5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9/conaninfo.txt
    assert "o=Y" in pkgb/2.52.0:5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9/conaninfo.txt
```

**Comment by memsharded** [comment]
```python
def test_private_error_options():
    # https://github.com/conan-io/conan/issues/9690
    c = TestClient()
    c.save({"pkgb/conanfile.py": GenConanfile().with_option("option", ["X", "Y"])
                                               .with_default_option("option", '"Y"'),
            "pkga/conanfile.py": GenConanfile().with_require("pkgb/1.0", private=True)
                                               .with_default_option("pkgb:option", '"X"'),
            "pkgm/conanfile.py": GenConanfile().with_require("pkgb/2.0", private=True)
                                               .with_require("pkga/0.1")
           })
    c.run("create pkgb pkgb/1.0@ -o pkgb:option=X")
    c.run("create pkgb pkgb/2.0@ -o pkgb:option=Y")
    c.run("search pkgb/1.0@")
    assert "08ddf75242c479442c18c7b053eccf9f9c70af0a" in c.out
    assert "option: X" in c.out
    c.run("search pkgb/2.0@")
    assert "b39f53a12b7d16e2ab23e6484c9e15268e7a77b2" in c.out
    assert "option: Y" in c.out
    c.run("create pkga pkga/0.1@")
    c.run("create pkgm pkgm/0.1@")
    assert "pkgb/1.0 from local cache - Cache" in c.out
    assert "pkgb/2.0 from local cache - Cache" in c.out
    assert "pkgb/1.0:08ddf75242c479442c18c7b053eccf9f9c70af0a - Skip" in c.out
    assert "pkgb/2.0:b39f53a12b7d16e2ab23e6484c9e15268e7a77b2 - Cache" in c.out

    c.run("create pkgm pkgm/0.1@ --build")
    assert "pkgb/1.0 from local cache - Cache"
    assert "pkgb/2.0 from local cache - Cache"
    assert "pkgb/1.0:08ddf75242c479442c18c7b053eccf9f9c70af0a - Skip" not in c.out
    assert "pkgb/2.0:b39f53a12b7d16e2ab23e6484c9e15268e7a77b2 - Build" in c.out
```
This works. The key point here is that you cannot define ``pkgb:option=Y`` in ``pkgm``, because that will produce an override of option values upstream. In the test I am producing different options using the default one, and the change is done in ``pkga``. It could be argued that definitions of options upstream could be truncated in transitive private, but the thing is that the ``private`` requires in Conan 1.X is complex, and changing anything would have a high risk of breaking other users, so better wait until 2.0 for this. If it is not possible to use the ``pkgb`` default (it should with some consensus), then it is necessary to create an intermediate ``pkgc``, that would be the one requiring ``pkga`` privately and setting the option value.

**Comment by sbellus** [comment]
Thanks for explanation. I will wait for conan 2.x.

**Comment by memsharded** [comment]
Closing this as solved by 2.0
